docs/site/Parsing-requests.md

```diff
@@ -234,6 +234,76 @@ class Product extends Entity {
 A full list of options & usage scenarios could be found in the
 [documentation of AJV errors](https://github.com/epoberezkin/ajv-errors).
 
+##### Custom AJV Formats
+
+AJV Custom Formats can be configured by binding the value
+`validation.ajvCustomFormats` to `RestBindings.REQUEST_BODY_PARSER_OPTIONS`
+('rest.requestBodyParserOptions').
+
+For example,
+
+```ts
+server.bind(RestBindings.REQUEST_BODY_PARSER_OPTIONS).to({
+  validation: {
+    ajvCustomFormats: {
+      'only-alphanumeric': /^[a-zA-Z0-9 ]+$/.source,
+      'product-code': {
+        async: true,
+        validate: async (code: string) => {
+          return /^[a-zA-Z]{3}-\d{3}$/.test(code)
+            ? Promise.resolve(true)
+            : Promise.resolve(false);
+        },
+      },
+    },
+  },
+});
+```
+
+A full list of options for adding custom ajv formats could be found in the
+[documentation of AJV add format](https://ajv.js.org/#api-addformat).
+
+##### Custom AJV Keywords
+
+AJV Custom Keywords can be configured by binding the value
+`validation.ajvCustomKeywords` to `RestBindings.REQUEST_BODY_PARSER_OPTIONS`
+('rest.requestBodyParserOptions').
+
+```ts
+server.bind(RestBindings.REQUEST_BODY_PARSER_OPTIONS).to({
+  validation: {
+    ajvCustomFormats: {
+      'only-alphanumeric': /^[a-zA-Z0-9 ]+$/.source,
```

Comment:
As per-your earlier comment, AJV custom formats can be useful for using functions for validation. I think we should prevent strings or regexp from being used as that'd overlap with the use of the `pattern` key.

Reply:
I do not see the reason to restrict string or regexp. Let the programmer trying to implement custom format decide on best use either via string/regexp in custom format or creating own logic for reusing string/regexp.

Reply:
_If_ we were to go forward with the duplicate RegExp validation feature, I see 2 concerns:

1. Will `ajvCustomFormats` be reflected on the [OAS 3 Schema Object Property, `pattern`](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.2.md#properties)?
2. If both a RegExp `format` and `pattern` are defined, which takes priority (for validation and OAS 3 `pattern` property)?

IMO while it's good to give developers options, duplicate features are not beneficial.

I think we need to more carefully-consider how we are exposing this API to the developers.

Reply:
Valid points.
1. Nope, `ajvCustomFormats` won't reflect on OAS 3 Schema property `pattern`.
2. I searched, however, I do not have an answer to which one takes priority between `format` and `pattern`. I will try running them in parallel with different regex to check which one is validated first.

`format` is a valid attribute of json schema validation draft though. http://json-schema.org/draft/2019-09/json-schema-validation.html#rfc.section.7

Reply:
1. If we were to allow RegExp `ajvCustomFormats`, it should update OAS 3 `pattern`.
2. As per-my previous comment, if we're going to implement this, I think we should restrict to function-based validation only to prevent this overlap. Taking a shot in the dark, both Regexp `ajvCustomFormats` and `pattern` will be used for validation.

`format` is valid and is being used by lb4 for types such as `Date` (`date-time`). In this case, there's the same concern of priority (of which should be documented).